src/vifm.1
.TH vifm 1 "October 12, 2011" "" "Vifm"
.\" ---------------------------------------------------------------------------
.SH NAME
.\" ---------------------------------------------------------------------------
vifm [OPTION]
.\" ---------------------------------------------------------------------------
.SH DESCRIPTION
.\" ---------------------------------------------------------------------------
Vifm is a ncurses based file manager with vi like keybindings. If you use vi,
vifm gives you complete keyboard control over your files without having to learn
a new set of commands.
.\" ---------------------------------------------------------------------------
.SH OPTIONS
.\" ---------------------------------------------------------------------------
The vifm executable will start vifm in the current directory unless it is
given a different directory on the command line.
.TP
.BI <lwinpath>
Starts Vifm in the specified path.
.TP
.BI "<lwinpath> <rwinpath>"
Starts Vifm in the specified paths.
.LP
Paths to files are also allowed in case you want vifm to start with some
archive opened.
.TP
.BI -f
only used from the vifm.vim script.  The selected files are written
to CONFDIR/vimfiles and vifm exits.
.TP
.BI "--logging"
Log some errors to CONFDIR/log.
.TP
.BI "--help, -h"
Show an overview of the commandline options.
.TP
.BI "--version, -v"
Show version information and quit.
.TP
.BI --no-configs
Don't read vifmrc and vifminfo.

.LP
See Configure section below for the explanations on CONFDIR.
.br
.\" ---------------------------------------------------------------------------
.SH General keys
.\" ---------------------------------------------------------------------------
.TP
.BI "Ctrl-c or Escape"
cancel most operations.
.TP
.BI "Ctrl-c or Escape"
clear all selected files.
.TP
.BI Ctrl-l
clear and redraw the screen.
.\" ---------------------------------------------------------------------------
.SH Basic Movement
.\" ---------------------------------------------------------------------------
The basic vi key bindings are used to move through the files and popup windows.
.TP
.BI "k or Ctrl-P"
moves up.
.TP
.BI "j or Ctrl-N"
moves down.
.TP
.BI h
moves up one directory.
.TP
.BI "l or Enter"
moves into a directory or launches a file.
.TP
.BI gg
move to the top of the file list.
.TP
.BI G
move to the bottom of the file list.
.TP
.BI H
move to the first file in the window.
.TP
.BI M
move to the file in the middle of the window.
.TP
.BI L
move to the last file in the window.
.TP
.BI "Ctrl-F or Page Down"
move forward one page.
.TP
.BI "Ctrl-B or Page Up"
move back one page.
.TP
.BI Ctrl-D
jump back one half page.
.TP
.BI Ctrl-U
jump forward one half page.
.TP
.BI n%
move to the file that is n percent from the top of the list (for example 25%).
.TP
.BI "Space Bar or Tab Key"
toggles between the two file lists.
.\" ---------------------------------------------------------------------------
.SH Movement with Count
.\" ---------------------------------------------------------------------------
.TP
Most movement commands also accept a count, 12j would move down 12 files.
.TP
.BI [count]%
move to percent of the file list.
.TP
.BI [count]j
move down count files.
.TP
.BI [count]k
move up count files.
.TP
.BI "[count]G or [count]gg"
move to list position count.
.\" ---------------------------------------------------------------------------
.SH Scrolling panes
.\" ---------------------------------------------------------------------------
.TP
.BI zt
redraw pane with file in top of list.
.TP
.BI zz
redraw pane with file in center of list.
.TP
.BI zb
redraw pane with file in bottom of list.
.TP
.BI Ctrl-E
scroll pane one line down.
.TP
.BI Ctrl-Y
scroll pane one line up.
.\" ---------------------------------------------------------------------------
.SH Pane manipulation
.\" ---------------------------------------------------------------------------
Second character can be entered with or without Control key.
.TP
.BI "Ctrl-W H"
switch to the left pane.
.TP
.BI "Ctrl-W L"
switch to the right pane.
.TP
.BI "Ctrl-W O"
shortcut for :only
.TP
.BI "Ctrl-W S"
shortcut for :split
.TP
.BI "Ctrl-W V"
shortcut for :split
.TP
.BI "Ctrl-W W"
switch to other pane.
.TP
.BI "Ctrl-W X"
exchange panes.
.\" ---------------------------------------------------------------------------
.SH Marks
.\" ---------------------------------------------------------------------------
.TP
Marks are set the same way as they are in vi.
.br
You can use this characters for marks [a-z][A-Z][0-9].
.TP
.BI m[a-z][A-Z][0-9]
to set a mark for the current file.
.TP
.BI '[a-z][A-Z][0-9]
moves to the file set for the mark.
.\" ---------------------------------------------------------------------------
.SH Searching
.\" ---------------------------------------------------------------------------
.TP
.BI "/regular expression pattern[Return]"
will highlight all files matching the pattern and go to the next match.
.TP
.BI "?regular expression pattern[Return]"
will highlight all files matching the pattern and go to the previous match.
.TP
.BI [count]n
find the next match of / or ?.
.TP
.BI [count]N
find the previous match of / or ?.
.TP
.BI [count]f[character]
search forward for file with [character] as first character in name.  Search
wraps around the end of the list.
.TP
.BI [count]F[character]
search backward for file with [character] as first character in name.  Search
wraps around the end of the list.
.TP
.BI [count];
find the next match of f or F.
.TP
.BI [count],
find the previous match of f or F.
.LP
Note: f, F, ; and , wrap around list beginning and end when they are used
alone and they don't wrap when they are used as selectors.
.\" ---------------------------------------------------------------------------
.SH File Filters
.\" ---------------------------------------------------------------------------
The basic vim folding key bindings are used for filtering files.
.TP
All directorys will show the ../ file no matter what the filter setting is with the exception of the / directory.
.TP
Each file list has its own filter.
.TP
Filtered files are not checked in / search or :commands.
.TP
.BI zo
Show all of the dot files.
.TP
.BI zf
Filter all of the selected files.
.TP
.BI za
Toggle the showing and hiding of dot files.
.TP
.BI zm
Filter all of the dot files.
.TP
.BI zO
Show the filtered files.
.TP
.BI zM
Filter the files matching the filename filter.
.TP
.BI zR
Remove all filters.
.\" ---------------------------------------------------------------------------
.SH Other Normal Mode Keys
.\" ---------------------------------------------------------------------------
.TP
.BI [count]:
enter command line mode.  Count will add range.
.TP
.BI "[count]!! and [count]!<selector>"
enter command line mode with entered ! command.  Count will modify range.
.TP
.BI Ctrl-O
go backward through history.
.TP
.BI Ctrl-I
go forward through history.
.TP
.BI Ctrl-G
creates a window showing detailed information about the current file.
.TP
.BI ga
calculate directory size.
.TP
.BI gA
like ga, but force update.
.TP
.BI gf
find link destination (like l with 'followlinks' off, but also finds
directories).
.TP
.BI gl
only for MS-Windows
.br
same as l key, but tries to run program with administrative privileges.
.TP
.BI gv
go to visual mode restoring last selection.
.TP
.BI gs
restore last t selection, like gv for visual mode selection.
.TP
.BI gu<selector>
make names of selected files lowercase.
.TP
.BI guu
.BI gugu
make name of current file lowercase.
.TP
.BI gU<selector>
make names of selected files uppercase.
.TP
.BI gUU
.BI gUgU
make name of current file uppercase.
.TP
.BI i
opens file with associated program even if it's an executable.
.TP
.BI cw
rename a file.
.TP
.BI cW
change only name of file (without extension).
.TP
.BI cl
change link target.
.TP
.BI co
only for *nix
.br
change file owner.
.TP
.BI cg
only for *nix
.br
change file group.
.TP
.BI cp
only for *nix
.br
change file permissions.
.TP
.BI C
clone file.
.TP
.BI "[count]dd or d[count]selector"
moves the selected files to the trash directory (if option 'trash' is set
otherwise delete)
.TP
.BI "[count]DD or D[count]selector"
removes the selected files.
.TP
.BI "Y, [count]yy or y[count]selector"
yanks the selected files.
.TP
.BI Y
same as yy.
.TP
.BI p
will copy the yanked files to the current directory or move the files
to the current directory if they were deleted with dd or :d[elete] or if the
files were yanked from the Trash directory.
.TP
.BI P
moves the last yanked files.  The advantage of using P instead of d followed by
p is that P moves files only once.  This isn't important in case you're moving
files in the same file system where your home directory is, but using P to
move files on some other file system (or file systems, in case you want to move
files from fs1 to fs2 and your home is on fs3) can save your time.
.TP
.BI al
puts symbolic links with absolute paths.
.TP
.BI rl
puts symbolic links with relative paths.
.TP
.BI t
tag the current file.
.TP
.BI u
undo last change.
.TP
.BI Ctrl-R
redo last change.
.TP
.BI v
enter visual mode.
.TP
.BI V
enter visual mode.
.TP
.BI ZQ
same as :quit!
.TP
.BI ZZ
same as :quit
.TP
.BI .
repeat last command line command (not normal mode command).
.TP
.BI (
goto previous group.  Groups are defined by primary sorting key.  For name and
iname members of each group have same first letter, for all other sorting keys
vifm uses size, uid, ...
.TP
.BI )
goto next group.  See ( key description above.
.\" ---------------------------------------------------------------------------
.SH Using Count
.\" ---------------------------------------------------------------------------
.TP
You can use count with commands like yy.
.TP
.BI [count]yy
yank count files starting from current cursor position downward.
.TP
Or you can use count with motions passed to y, d or D.
.TP
.BI d[count]j
delete (count + 1) files starting from current cursor position upward.
.\" ---------------------------------------------------------------------------
.SH Registers
.\" ---------------------------------------------------------------------------
.LP
vifm supports multiple registers for temporary storing list of yanked or
deleted files.

Registers should be specified with hitting double quite key followed by a
register name.  Count is specified after register name.  By default commands use
unnamed register, which has double quote as its name.

Though all commands accept registers, most of commands ignores them (for example
H or Ctrl-U).  Other commands can fill register or append new files to it.

Presently vifm supports ", _, a-z and A-Z characters as register names.

As mentioned above " is unnamed register and has special meaning of the default
register.  Every time when you use named registers (a-z and A-Z) unnamed
register is updated to contain same list of files as the last used register.

_ is black hole register.  It can be used for writing, but its list is always
empty.

Registers with names from a to z  and from A to Z are named ones.  Lowercase
registers are cleared before adding new files, while uppercase aren't and should
be used to append new files to the existing file list of appropriate lowercase
register (A for a, B for b, ...).

Registers can be changed on :empty command if they contain files under Trash
directory.

Registers do not contain one file more than once.

Example:
  "a2yy
.br
will put names of two files to register a (and to the unnamed register).
  "Ad
.br
will remove one file and append its name to register a (and to the unnamed
register).
  p or "ap or "Ap
.br
will insert previously yanked and deleted files into current directory.
.\" ---------------------------------------------------------------------------
.SH Selectors
.\" ---------------------------------------------------------------------------
.TP
y and d commands accept selectors.  You can combine them with every of
selectors below to quickly remove or yank several files.
.br
Most of selectors are like vi motions: j, k, gg, G, H, L, M, %, f, F, ;, comma
and '.  But there are some additional ones.
.TP
.BI a
all files in current view.
.TP
.BI s
selected files.
.TP
.BI S
all files except selected.
.\" ---------------------------------------------------------------------------
.SH Visual Mode
.\" ---------------------------------------------------------------------------
In visual mode work almost all normal mode keys, but they do not accept
selectors.
.TP
.BI Enter
save selection and go back to normal mode.
.TP
.BI gv
restore previous visual selection.
.TP
.BI v
leave visual mode.
.TP
.BI V
leave visual mode.
.TP
.BI :
enter command line mode.  When you leave it selection will be cleared.
.TP
.BI o
switch active selection bound.
.TP
.BI O
switch active selection bound.
.TP
.BI "gu, u"
make names of selected files lowercase.
.TP
.BI "gU, U"
make names of selected files uppercase.
.\" ---------------------------------------------------------------------------
.SH Command line Mode
.\" ---------------------------------------------------------------------------
This keys apply to all submodes of the command line mode: command, prompt and
search.
.LP
Down, Up, Left, Right, Home, End and Delete are extended keys and they are not
available if vifm is compiled with --disable-extended-keys option
.TP
.BI "Esc, Ctrl-C"
leave command line mode, cancels input.
.TP
.BI "Ctrl-M, Enter"
execute command and leave command line mode.
.TP
.BI "Ctrl-I, Tab"
complete command or its argument.
.TP
.BI Shift-Tab
complete in reverse order.
.TP
.BI Ctrl-_
stop completion and return original input.
.TP
.BI Ctrl-K
remove characters from cursor position till the end of line.
.TP
.BI Ctrl-U
remove characters from cursor position till the beginning of line.
.TP
.BI "Ctrl-H, Backspace"
remove character before the cursor.
.TP
.BI "Ctrl-D, Delete"
remove character under the cursor.
.TP
.BI "Ctrl-B, Left"
move cursor to the left.
.TP
.BI "Ctrl-F, Right"
move cursor to the right.
.TP
.BI "Ctrl-A, Home"
go to line beginning.
.TP
.BI "Ctrl-E, End"
go to line end
.TP
.BI Alt-B
go to the beginning of previous word.
.TP
.BI Alt-F
go to the end of next word.
.TP
.BI Ctrl-W
remove characters from cursor position till the beginning of previous word.
.TP
.BI Alt-D
remove characters from cursor position till the beginning of next word.
.TP
.BI Ctrl-N
recall more recent command-line from history.
.TP
.BI Ctrl-P
recall older command-line from history.
.TP
.BI Up
recall more recent command-line from history, that begins as the current
command-line.
.TP
.BI Down
recall older command-line from history, that begins as the current command-line.
.\" ---------------------------------------------------------------------------
.SH Commands
.\" ---------------------------------------------------------------------------
Commands are executed with :command_name<Return>

\'|' can be used to separate commands, so you can give multiple commands in one
line.  If you want to use '|' in an argument, precede it with '\\'.

Commented out lines should start with the double quote symbol, which may be
preceded by whitespace characters.
.TP
.BI :[count]
.TP
.BI :number
move to the file number.
.br
:12 would move to the 12th file in the list.
.br
:0 move to the top of the list.
.br
:$ move to the bottom of the list.
.TP
.BI :[count]command
The only builtin :[count]command are :[count]d[elete] and :[count]y[ank].
.TP
.BI :d3
would delete three files starting at the current file position moving down.
.TP
.BI :3d
would delete one file at the third line in the list.
.TP
.BI ":command [args]"
.TP
.BI :[range]!program
will execute the program in a shell
.TP
.BI ":[range]!command &"
.LP
will run the process in the background using vifm's means.

Programs that write to stdout like ls will create an error message
showing partial output of the command.

Take note of the space before ampersand symbol, if you omit it, command
will be run in the background using job control of your shell.

Accepts macros.
.TP
.BI ":[range]!! <program>"
is the same as :! but will pause the screen before returning to Vifm.
.TP
.BI :!!
will execute the last command.
.TP
.BI :[range]alink[!?]
creates absolute symbolic links of files in directory of other view.  With "?"
vifm will open vi to edit filenames.  "!" forces overwrite.
.TP
.BI ":[range]alink[!] path"
creates absolute symbolic links of files in directory specified with the path
(absolute or relative to directory of other view).  "!" forces overwrite.
.TP
.BI ":[range]alink[!] name1 name2..."
creates absolute symbolic links of files in directory of other view giving each
next link a corresponding name from the argument list.  "!" forces overwrite.
.TP
.BI ":apropos manpage"
will create a menu of items returned by the apropos command. Selecting an item
in the menu will open the corresponding manpage.
.TP
.BI :apropos
repeats last :apropos command.
.TP
.BI :cd
change to your home directory.
.TP
.BI ":cd -"
go to previous directory.
.TP
.BI ":cd ~/dir"
change directory to ~/dir.
.TP
.BI ":cd /curr/dir /other/dir"
change directory of the current pane to /curr/dir and directory of the other
pane to /other/dir.  When using relative paths vifm assumes that both of them
are relative to current directory of current view.  Command will not fail if one
of directories is invalid.
Accepts macros.
.TP
.BI ":cd! /dir"
same as :cd /dir /dir.
.TP
.BI :c[hange]
create a menu window to alter a files properties.
.TP
.BI :[range]chmod
only for *nix
.br
shows permission change dialog.
.TP
.BI ":[range]chmod[!] arg..."
only for *nix
.br
changes permissions for files.  See 'man chmod' for arg format.  "!" means set
permissions recursively.
.TP
.BI :[range]chown
only for *nix
.br
same as co key in normal mode.
.TP
.BI ":[range]chown [user][:][group]"
only for *nix
.br
changes owner and/or group of files.  Operates on directories recursively.
.TP
.BI :[range]clone[!?]
clones files in current directory.  With "?" vifm will open vi to edit
filenames.  "!" forces overwrite.
.TP
.BI ":[range]clone[!] path"
clones files to directory specified with the path (absolute or relative to
current directory).  "!" forces overwrite.
.TP
.BI ":[range]clone[!] name1 name2..."
clones files in current directory giving each next clone a corresponding name
from the argument list.  "!" forces overwrite.
.TP
.BI :colo[rscheme]?
prints current color scheme name in the status bar.
.TP
.BI :colo[rscheme]
gives a menu with a list of available color schemes.  You can choose default
color scheme here.  It will be used for view if no DIRECTORY in colorscheme
file fits current path.  It's also used to set border color (except view titles)
and colors in the menus and dialogs.
.TP
.BI ":colo[rscheme] color_scheme_name"
changes default color scheme to color_scheme_name.
.TP
.BI ":colo[rscheme] color_scheme_name directory"
associates directory with the color scheme.
.TP
.BI :comc[lear]
removes all user defined commands.
.TP
.BI :com[mand]
gives a menu of user commands.
.TP
.BI ":com[mand] beginning"
shows user defined commands that start with the beginning.
.TP
.BI ":com[mand] name action"
sets a new user command.
.br
Trying to use a reserved command name will result in an error message.
.br
Use :com[mand]! to overwrite a previously set command.
.br
Unlike vim user commands do not have to start with a capital letter.
User commands are run in a shell by default.  To run a command in
the background you must set it as a background command with & at the end
of the commands action (:com rm rm %f &).  Command name cannot contain
numbers or special symbols (except '?' and '!').
.TP
.BI ":com[mand] backup /pattern"
will set search pattern.
.TP
.BI ":com[mand] name filter pattern"
will set file name filter.
.TP
.BI ":[range]co[py][!?][ &]"
copies files to directory of other view.  With "?" vifm will open vi to edit
filenames.  "!" forces overwrite.
.TP
.BI ":[range]co[py][!] path[ &]"
copies files to directory specified with the path (absolute or relative to
directory of other view).  "!" forces overwrite.
.TP
.BI ":[range]co[py][!] name1 name2...[ &]"
copies files to directory of other view giving each next file a corresponding
name from the argument list.  "!" forces overwrite.
.TP
.BI ":[range]d[elete][!][ &]"
delete selected file or files.  "!" means completely remove file.
.TP
.BI ":[range]d[elete][!] [reg] [count][ &]"
will delete files to the reg register.  "!" means completely remove file.
.TP
.BI ":delc[ommand] command_name"
will remove the command_name user command.
.TP
.BI :delm[arks]!
will delete all marks.
.TP
.BI ":delm[arks] marks ..."
will delete specified marks, each argument is treated as a set of marks.
.TP
.BI :di[splay]
popup menu with registers content.
.TP
.BI ":di[splay] list ..."
display the contents of the numbered and named registers that are mentioned in
list (for example "az to display "", "a and "z content).
.TP
.BI :dirs
display directory stack.
.TP
.BI ":[range]e[dit] [file...]"
will load the selected or passed file or files into vi.  Accepts macros.
.TP
.BI :empty
will permanently remove 'rm -fr' files from the Trash directory.  It will also
remove all operations from undolist that have no sense after :empty and remove
all records about files in the Trash directory from all registers.
.TP
.BI :exi[t][!]
same as :quit.
.TP
.BI :f[ile]
popup menu of programs set for the file type of the current file. Add ' &' at
the end of command to run program in background.
.TP
.BI ":f[ile] arg"
run associated command that begins with the arg without opening menu.
.TP

NEWS
v0.6.0:
    Added:
    - ZZ and ZE commands
    - Ctrl+D and Ctrl+U commands
    - zb, zt and zz commands
    - Ctrl+E and Ctrl+Y commands
    - Ctrl+W W, Ctrl+W O and Ctrl+W V commands
    - i command to force file opening even for executables
    - IEC abbreviations
    - Ctrl+U, Ctrl+N, Ctrl+P, Ctrl+K, Ctrl+W, Meta+F and Meta+B in command line
      mode
    - Home and End keys support in command line mode
    - backward search (?)
    - :map, :nmap:, :cmap, :vmap commands
    - co (change owner) and cg (change group) commands
    - utf8 support
    - support for motions with d and y commands

    Changed:
    - keys input handling
    - force dot files appear at the beginning of file list
    - removed limit command line length limit
    - history of visited directories
    - ascending/descending sort for all types of criteria

    Fixed:
    - segmentation fault caused by newwin
    - segmentation fault caused by bookmarks menu
    - now view is updated after several commands
    - bug with empty command line
    - file renaming
    - ranges handling
    - behavior of n and N commands for different search directions
    - filename completion
    - user commands completion
    - memory allocation for history
    - work with registers


src/vifm.c
/* vifm
 * Copyright (C) 2001 Ken Steen.
 * Copyright (C) 2011 xaizek.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#include "../config.h"

#ifdef _WIN32
#include <tchar.h>
#include <windows.h>
#endif

#ifdef HAVE_LIBGTK
#include <glib-2.0/gio/gio.h>
#include <gtk/gtk.h>
#endif

#include <curses.h>

#include <unistd.h> /* getcwd, stat, sysconf */

#include <locale.h> /* setlocale */
#include <string.h> /* strncpy */

#include "background.h"
#include "bookmarks.h"
#include "color_scheme.h"
#include "commands.h"
#include "config.h"
#include "filelist.h"
#include "fileops.h"
#include "filetype.h"
#include "log.h"
#include "macros.h"
#include "main_loop.h"
#include "menus.h"
#include "modes.h"
#include "normal.h"
#include "ops.h"
#include "opt_handlers.h"
#include "registers.h"
#include "signals.h"
#include "sort.h"
#include "status.h"
#include "tree.h"
#include "ui.h"
#include "undo.h"
#include "utils.h"

#ifndef _WIN32
#define CONF_DIR "~/.vifm"
#else
#define CONF_DIR "(%HOME%/.vifm or %APPDATA%/Vifm)"
#endif

static void
show_version_msg(void)
{
	int i, len;
	char **list;
	list = malloc(sizeof(char*)*fill_version_info(NULL));

	len = fill_version_info(list);
	for(i = 0; i < len; i++)
	{
		puts(list[i]);
		free(list[i]);
	}

	free(list);
}

static void
show_help_msg(void)
{
	puts("vifm usage:\n");
	puts("  To start in a specific directory give the directory path.\n");
	puts("    vifm /path/to/start/dir/one");
	puts("    or");
	puts("    vifm /path/to/start/dir/one  /path/to/start/dir/two\n");
	puts("  To open file using associated program pass to vifm it's path.\n");
	puts("  If no path is given vifm will start in the current working directory.\n");
	puts("  vifm --logging");
	puts("    log some errors to " CONF_DIR "/log.\n");
	puts("  vifm --version | -v");
	puts("    show version number and quit.\n");
	puts("  vifm --help | -h");
	puts("    show this help message and quit.\n");
	puts("  vifm --no-configs");
	puts("    don't read vifmrc and vifminfo.");
}

static void
init_window(FileView *win)
{
	win->curr_line = 0;
	win->top_line = 0;
	win->list_rows = 0;
	win->list_pos = 0;
	win->selected_filelist = NULL;
	win->history_num = 0;
	win->history_pos = 0;
	win->invert = 0;
	win->color_scheme = 1;
}

static void
init_window_history(FileView *win)
{
	if(cfg.history_len == 0)
		return;

	win->history = malloc(sizeof(history_t)*cfg.history_len);
	while(win->history == NULL)
	{
		cfg.history_len /= 2;
		win->history = malloc(sizeof(history_t)*cfg.history_len);
	}
}

static void
load_initial_directory(FileView *view, const char *dir)
{
	if(view->curr_dir[0] == '\0')
		snprintf(view->curr_dir, sizeof(view->curr_dir), "%s", dir);
	else
		dir = view->curr_dir;

	view->dir_entry = (dir_entry_t *)malloc(sizeof(dir_entry_t));
	memset(view->dir_entry, 0, sizeof(dir_entry_t));

	view->dir_entry[0].name = malloc(sizeof("../") + 1);
	strcpy(view->dir_entry[0].name, "../");
	view->dir_entry[0].type = DIRECTORY;

	view->list_rows = 1;
	chosp(view->curr_dir);
	(void)change_directory(view, dir);
}

/* buf should be at least PATH_MAX characters length */
static void
parse_path(const char *dir, const char *path, char *buf)
{
	if(is_path_absolute(path))
	{
		snprintf(buf, PATH_MAX, "%s", path);
	}
#ifdef _WIN32
	else if(path[0] == '/')
	{
		snprintf(buf, PATH_MAX, "%c:%s", dir[0], path);
	}
#endif
	else
	{
		char new_path[PATH_MAX];
		snprintf(new_path, sizeof(new_path), "%s/%s", dir, path);
		canonicalize_path(new_path, buf, PATH_MAX);
	}
	if(!is_root_dir(buf))
		chosp(buf);
}

static void
parse_args(int argc, char *argv[], const char *dir, char *lwin_path,
		char *rwin_path)
{
	int x;

	(void)my_chdir(dir);

	/* Get Command Line Arguments */
	for(x = 1; x < argc; x++)
	{
		if(argv[x] == NULL)
		{
			continue;
		}
		else if(!strcmp(argv[x], "-f"))
		{
			cfg.vim_filter = 1;
		}
		else if(!strcmp(argv[x], "--no-configs"))
		{
			continue;
		}
		else if(!strcmp(argv[x], "--version") || !strcmp(argv[x], "-v"))
		{
			endwin();
			show_version_msg();
			exit(0);
		}
		else if(!strcmp(argv[x], "--help") || !strcmp(argv[x], "-h"))
		{
			endwin();
			show_help_msg();
			exit(0);
		}
		else if(!strcmp(argv[x], "--logging"))
		{
			init_logger(1);
		}
		else if(access(argv[x], F_OK) == 0 || is_path_absolute(argv[x]) ||
				is_root_dir(argv[x]))
		{
			if(lwin_path[0] != '\0')
				parse_path(dir, argv[x], rwin_path);
			else
				parse_path(dir, argv[x], lwin_path);
		}
		else
		{
			endwin();
			show_help_msg();
			exit(0);
		}
	}
}

static void
update_path(void)
{
#ifndef _WIN32
	char *old_path;
	char *new_path;

	old_path = getenv("PATH");
	new_path = malloc(strlen(cfg.config_dir) + 8 + 1 + strlen(old_path) + 1);
	sprintf(new_path, "%s/scripts:%s", cfg.config_dir, old_path);
	setenv("PATH", new_path, 1);
	free(new_path);
#else
	char *old_path;
	char *new_path;
	int i;

	old_path = getenv("PATH");
	new_path = malloc(5 + strlen(cfg.config_dir) + 8 + 1 + strlen(old_path) + 1);
	sprintf(new_path, "PATH=%s/scripts;%s", cfg.config_dir, old_path);

	for(i = 0; new_path[i] != '\0'; i++)
	{
		if(new_path[i] == '/')
			new_path[i] = '\\';
	}

	putenv(new_path);
	free(new_path);
#endif
}

static void
check_path(FileView *view, const char *path)
{
	if(path[0] == '\0')
		return;

	strcpy(view->curr_dir, path);
	if(!is_dir(path) && !is_unc_root(path))
	{
		char *slash;
		if((slash = strrchr(view->curr_dir, '/')) != NULL)
			*slash = '\0';
	}
}

static void
check_path_for_file(FileView *view, const char *path)
{
	load_dir_list(view, !(cfg.vifm_info&VIFMINFO_SAVEDIRS));
	if(path[0] != '\0' && !is_dir(path))
	{
		char *slash = strrchr(path, '/');
		int pos;
		if(slash != NULL && (pos = find_file_pos_in_list(view, slash + 1)) >= 0)
		{
			view->list_pos = pos;
			handle_file(view, 0, 0);
		}
	}
}

static int
run_converter(int vifm_like)
{
#ifndef _WIN32
	char buf[PATH_MAX];
	snprintf(buf, sizeof(buf), "vifmrc-converter %d", vifm_like);
	return shellout(buf, -1);
#else
	TCHAR buf[PATH_MAX + 2];

	if(GetModuleFileName(NULL, buf, ARRAY_LEN(buf)) == 0)
		return -1;

	*(_tcsrchr(buf, _T('\\')) + 1) = _T('\0');
	if(vifm_like == 2)
		_tcscat(buf, _T("vifmrc-converter 2"));
	else if(vifm_like == 1)
		_tcscat(buf, _T("vifmrc-converter 1"));
	else
		_tcscat(buf, _T("vifmrc-converter 0"));

	return exec_program(buf);
#endif
}

int
main(int argc, char *argv[])
{
	char dir[PATH_MAX];
	char config_dir[PATH_MAX];
	char *console = NULL;
	char lwin_path[PATH_MAX] = "";
	char rwin_path[PATH_MAX] = "";
	int old_config;
	int i;
	int no_configs;

	(void)setlocale(LC_ALL, "");
	if(getcwd(dir, sizeof(dir)) == NULL)
	{
		perror("getcwd");
		return -1;
	}
#ifdef _WIN32
	to_forward_slash(dir);
#endif

	init_window(&rwin);
	init_window(&lwin);

	filetypes = NULL;

	init_registers();
	init_config();
	set_config_dir();

	update_path();

	init_commands();
	load_default_configuration();

	/* Safety check for existing vifmrc file without FUSE_HOME */
	if(cfg.fuse_home == NULL)
		cfg.fuse_home = strdup("/tmp/vifm_FUSE");

	/* Misc configuration */

	lwin.prev_invert = lwin.invert;
	lwin.hide_dot = 1;
	strncpy(lwin.regexp, "", sizeof(lwin.regexp));
	lwin.matches = 0;
	init_window_history(&lwin);

	rwin.prev_invert = rwin.invert;
	rwin.hide_dot = 1;
	strncpy(rwin.regexp, "", sizeof(rwin.regexp));
	rwin.matches = 0;
	init_window_history(&rwin);

	init_status();
	curr_stats.dirsize_cache = tree_create(0, 0);
	if(curr_stats.dirsize_cache == NULL)
	{
		puts("Not enough memory for initialization");
		return -1;
	}

#ifdef HAVE_LIBGTK
	curr_stats.gtk_available = gtk_init_check(&argc, &argv);
#endif

	if(cfg.show_one_window)
		curr_stats.number_of_windows = 1;
	else
		curr_stats.number_of_windows = 2;

	snprintf(config_dir, sizeof(config_dir), "%s/vifmrc", cfg.config_dir);

	/* Check if running in X */
#ifndef _WIN32
	console = getenv("DISPLAY");
#else
	console = "WIN";
#endif
	if(!console || !*console)
		curr_stats.is_console = 1;

	curr_view = &lwin;
	other_view = &rwin;

	/* Setup the ncurses interface. */
	if(!setup_ncurses_interface())
		return -1;

	no_configs = 0;
	for(i = 1; i < argc; i++)
		if(strcmp("--no-configs", argv[i]) == 0)
			no_configs = 1;

	old_config = is_old_config();
	if(!old_config && !no_configs)
		read_info_file(0);

	parse_args(argc, argv, dir, lwin_path, rwin_path);
	check_path(&lwin, lwin_path);
	check_path(&rwin, rwin_path);

	load_initial_directory(&lwin, dir);
	load_initial_directory(&rwin, dir);

	init_modes();
	init_option_handlers();
	init_undo_list(&perform_operation, &cfg.undo_levels);
	load_local_options(curr_view);

	curr_stats.vifm_started = 1;

	if(!old_config && !no_configs)
	{
		if(are_old_color_schemes())
		{
			if(run_converter(2) != 0)
			{
				endwin();
				fputs("Problems with running vifmrc-converter", stderr);
				exit(0);
			}
		}
		if(find_color_scheme(curr_stats.color_scheme))
			load_color_scheme(curr_stats.color_scheme);
		load_color_scheme_colors();
		exec_config();
	}

	write_color_scheme_file();
	setup_signals();

	if(old_config && !no_configs)
	{
		int vifm_like;
		int result;
		if(!query_user_menu("Configuration update", "Your vifmrc will be "
				"upgraded to a new format.  Your current configuration will be copied "
				"before performing any changes, but if you don't want to take the risk "
				"and would like to make one more copy say No to exit vifm.  Continue?"))
		{
#ifdef _WIN32
			system("cls");
#endif
			endwin();
			exit(0);
		}

		vifm_like = !query_user_menu("Configuration update", "This version of vifm "
				"is able to save changes in the configuration files automatically when "
				"quitting, as it was possible in older versions.  It is from now on "
				"recommended though, to save permanent changes manually in the "
				"configuration file as it is done in vi/vim.  Do you want vifm to "
				"behave like vi/vim?");

		result = run_converter(vifm_like);
		if(result != 0)
		{
			endwin();
			fputs("Problems with running vifmrc-converter", stderr);
			exit(0);
		}

		(void)show_error_msg("Configuration update", "Your vifmrc has been "
				"upgraded to "
				"new format, you can find its old version in " CONF_DIR "/vifmrc.bak.  "
				"vifm will not write anything to vifmrc, and all variables that are "
				"saved between runs of vifm are stored in " CONF_DIR "/vifminfo now "
				"(you can edit it by hand, but do it carefully).  You can control what "
				"vifm stores in vifminfo with 'vifminfo' option.");

		curr_stats.vifm_started = 0;
		read_info_file(0);
		curr_stats.vifm_started = 1;

		check_path(&lwin, lwin_path);
		check_path(&rwin, rwin_path);

		load_initial_directory(&lwin, dir);
		load_initial_directory(&rwin, dir);

		exec_config();
	}

	check_path_for_file(&lwin, lwin_path);
	check_path_for_file(&rwin, rwin_path);

	curr_stats.vifm_started = 2;
	modes_redraw();
	if(curr_stats.startup_redraw_pending)
		redraw_window();
	main_loop();

	return 0;
}

/* vim: set tabstop=2 softtabstop=2 shiftwidth=2 noexpandtab cinoptions-=(0 : */
/* vim: set cinoptions+=t0 : */
